diffhtwo/experimental/tests/test_diffstarpop_halpha_opt.py

```diff
@@ -0,0 +1,248 @@
+import numpy as np
+import jax.numpy as jnp
+from jax import random as jran
+from dsps.cosmology import DEFAULT_COSMOLOGY
+from dsps.data_loaders import retrieve_fake_fsps_data
+from jax.flatten_util import ravel_pytree
+from numbers import Real
+from diffstar.defaults import T_TABLE_MIN
+from diffstarpop.defaults import DEFAULT_DIFFSTARPOP_PARAMS
+from dsps.cosmology import flat_wcdm
+from dsps.metallicity import umzr
+from diffsky.experimental import mc_lightcone_halos as mclh
+from diffsky.param_utils import spspop_param_utils as spspu
+from ..diffstarpop_halpha import diffstarpop_halpha_kern as dpop_halpha_L
+from ..diffstarpop_halpha import (
+    diffstarpop_halpha_lf_weighted as dpop_halpha_lf_weighted,
+)
+from ..diffstarpop_halpha_opt import fit_diffstarpop
+
+# from diffsky.experimental.scatter import DEFAULT_SCATTER_PARAMS
+# from diffsky.ssp_err_model import ssp_err_model
+
+
+theta_default, unravel_fn = ravel_pytree(DEFAULT_DIFFSTARPOP_PARAMS)
+idx = jnp.array([1, 8, 16])
+
+
+def _jitter_value(v, s, rng):
+    if isinstance(v, Real):
+        eps = s * (abs(v) if v != 0 else 1.0)
+        return v + rng.normal(0.0, eps)
+    return v
+
+
+def perturb_namedtuple(nt, scale=0.01, rng=None, bounds=None, per_field_scale=None):
+    """
+    Perturb a namedtuple's numeric fields with small Gaussian noise.
+
+    - scale: default fractional std for all numeric fields
+    - per_field_scale: dict(field -> scale) to override per field
+    - bounds: dict(field -> (lo, hi)) to clamp after jitter
+    """
+    if rng is None:
+        rng = np.random.default_rng()
+
+    updates = {}
+    for f in nt._fields:
+        v = getattr(nt, f)
+
+        # Recurse if nested namedtuple
+        if hasattr(v, "_fields"):
+            updates[f] = perturb_namedtuple(
+                v, scale=scale, rng=rng, bounds=bounds, per_field_scale=per_field_scale
+            )
+            continue
+
+        s = per_field_scale.get(f, scale) if per_field_scale else scale
+        v2 = _jitter_value(v, s, rng)
+
+        if bounds and f in bounds:
+            lo, hi = bounds[f]
+            v2 = float(np.clip(v2, lo, hi))
+
+        updates[f] = v2
+
+    return nt._replace(**updates)
+
+
+def perturb_diffstarpop(params, scale=0.01, rng=None):
+    """
+    Convenience wrapper for your DiffstarPopParams-like namedtuple,
+    with sensible default bounds for probability-like params and stds.
+    """
+    # Heuristic bounds for common fields (edit to taste)
+    bnds = {}
+    # Anything that looks like a fraction or ylo/yhi: clamp to [0, 1]
```

Comment:
Ok, this is important: DiffstarPop actually has bounds defined for each parameter. The same is true for all diffsky model components. And the are not always square, and certainly not always between 0 and 1. We should really be running these gradient descents on the _unbounded_ parameters rather than the bounded ones.